README.md
# pytineye

**pytineye** is the official Python client library for the TinEye API. The TinEye API
is TinEye's paid search alternative for professional, commercial or high-volume users.
See <https://api.tineye.com/> for more information.

# Contents

- [pytineye](#pytineye)
- [Contents](#contents)
- [Installation](#installation)
- [Migrating from previous versions](#migrating-from-previous-versions)
      - [New ✅](#new-)
      - [Old ❌](#old-)
- [Documentation](#documentation)
  - [Building the documentation](#building-the-documentation)
- [Support](#support)

# Installation

Install the latest version of the library using pip:

```bash
$ python -m pip install pytineye
```

# Migrating from previous versions

If you were using any version of **pytineye** before `3.0.0`, you will need
to make minor changes to your code.

The API object is now instantiated using a single key, `api_key`. The value
of this key is the same as your previous `private_key`. The public key is no 
longer used.

#### New ✅
```python
# Sandbox key
# Note that this is the same value as the old private_key
api_key = "6mm60lsCNIB,FwOWjJqA80QZHh9BMwc-ber4u=t^"
api = TinEyeAPIRequest(
    api_url="https://api.tineye.com/rest/",
    api_key=api_key,
)
```

#### Old ❌
```python
# Sandbox keys
public_key = "LCkn,2K7osVwkX95K4Oy"
private_key = "6mm60lsCNIB,FwOWjJqA80QZHh9BMwc-ber4u=t^"
api = TinEyeAPIRequest(
    api_url="https://api.tineye.com/rest/",
    public_key=public_key,
    private_key=private_key,
)
```

# Documentation

View [documentation](https://api.tineye.com/python/docs/).

## Building the documentation

To build the documentation 

Install [sphinx](https://pypi.org/project/Sphinx/)
```
pip install sphinx
```

Run the sphinx-build command
```
sphinx-build docs/ docs/_build
```

# Support

Please send comments, recommendations, and bug reports to <support@tineye.com>.


CHANGELOG.md
Changelog
=========

3.1.0
-----
* Add read_timeout and connect_timeout to TinEyeAPIRequest

3.0.0
-----
* Swapping HMAC-SHA256 authentication with header key

2.0.1
-----
* Modifying boundary string so it starts and ends with `--` for multipart requests.

2.0.0
-----
* Fixed a bug in the construction of the string to sign.
* Added timeouts to TinEye API requests.
* Used [black](https://github.com/psf/black) to reformat code.
* Updated versions for dependencies.

1.3
---
* Removed `total_results` from TinEyeResponse object since it's been moved to `stats`.
* Removed `contributor` attribute from Match object and replacing it with `tags` list.
* Added `stats` attribute to TinEyeResponse which is a dictionary containing statistics about
  the search.
* Added `score` and `domain` attributes to TinEyeResponse.
* Updated to use SHA256 to sign requests to the TinEye API.
* Switched to using `urllib3[secure]`.

1.2
---
* Changing `remaining_searches` method to fit the new format.

1.1
---
* Added Python 3 support.
* Replaced `simplejson` with `json`.

1.0
---
* Added `total_results` to TinEyeResponse.

0.1
---
* Initial release.


test/test_api.py
# -*- coding: utf-8 -*-

"""
test_api_request.py

Test TinEyeAPIRequest class.

Copyright (c) 2021 TinEye. All rights reserved worldwide.
"""

import unittest
from datetime import datetime

from pytineye.api import Backlink, Match, TinEyeAPIRequest, TinEyeResponse


class TestTinEyeAPIRequest(unittest.TestCase):
    """ Test TinEyeAPIRequest class. """

    def setUp(self):
        self.api = TinEyeAPIRequest(
            api_url="https://api.tineye.com/rest/",
            api_key="6mm60lsCNIB,FwOWjJqA80QZHh9BMwc-ber4u=t^",
        )

    def test_backlink(self):
        """ Test TinEyeAPI.Backlink object. """

        backlink = {"url": "url", "crawl_date": "2010-02-19", "backlink": "backlink"}
        b = Backlink._from_dict(backlink)
        self.assertEqual(
            repr(b),
            'Backlink(url="url", backlink="backlink", crawl_date=2010-02-19 00:00:00)',
        )
        self.assertEqual(b.url, "url")
        self.assertEqual(b.crawl_date, datetime(2010, 2, 19, 0, 0))
        self.assertEqual(b.backlink, "backlink")

        backlink = {"url": "url", "crawl_date": "", "backlink": "backlink"}
        b = Backlink._from_dict(backlink)
        self.assertEqual(b.url, "url")
        self.assertEqual(b.crawl_date, datetime(1, 1, 1, 0, 0))
        self.assertEqual(b.backlink, "backlink")

        backlink = {"url": "", "crawl_date": "", "backlink": ""}
        b = Backlink._from_dict(backlink)
        self.assertEqual(b.url, "")
        self.assertEqual(b.crawl_date, datetime(1, 1, 1, 0, 0))
        self.assertEqual(b.backlink, "")

        backlink = {"url": None, "crawl_date": None, "backlink": None}
        b = Backlink._from_dict(backlink)
        self.assertEqual(b.url, None)
        self.assertEqual(b.crawl_date, datetime(1, 1, 1, 0, 0))
        self.assertEqual(b.backlink, None)

    def test_match(self):
        """ Test TinEyeAPI.Match object. """

        match = {
            "backlinks": [
                {"url": "url", "crawl_date": "2008-04-27", "backlink": "backlink"}
            ],
            "domain": "domain",
            "score": 14.8,
            "format": "JPEG",
            "overlay": "overlay",
            "height": 297,
            "width": 350,
            "image_url": "image_url",
            "filesize": 87918,
            "contributor": False,
            "size": 103950,
            "query_hash": "dca08fc6b2ec4b9e04f94a4e29223f6af3dd6555",
        }
        m = Match._from_dict(match)
        self.assertEqual(
            repr(m), 'Match(image_url="image_url", score=14.80, width=350, height=297)'
        )
        self.assertEqual(len(m.backlinks), 1)
        self.assertEqual(m.domain, "domain")
        self.assertEqual(m.score, 14.8)
        self.assertEqual(m.format, "JPEG")
        self.assertEqual(m.overlay, "overlay")
        self.assertEqual(m.height, 297)
        self.assertEqual(m.width, 350)
        self.assertEqual(m.image_url, "image_url")
        self.assertEqual(m.filesize, 87918)
        self.assertEqual(m.tags, [])
        self.assertEqual(m.size, 103950)

        match = {
            "backlinks": [
                {"url": "url", "crawl_date": "2008-04-27", "backlink": "backlink"},
                {"url": "url", "crawl_date": "2009-04-27", "backlink": "backlink"},
            ],
            "domain": "domain",
            "score": 67.19,
            "format": "JPEG",
            "overlay": "overlay",
            "height": 297,
            "width": 350,
            "image_url": "image_url",
            "filesize": 87918,
            "contributor": True,
            "size": 103950,
        }
        m = Match._from_dict(match)
        self.assertEqual(len(m.backlinks), 2)
        self.assertEqual(m.domain, "domain")
        self.assertEqual(m.score, 67.19)
        self.assertEqual(m.format, "JPEG")
        self.assertEqual(m.overlay, "overlay")
        self.assertEqual(m.height, 297)
        self.assertEqual(m.width, 350)
        self.assertEqual(m.image_url, "image_url")
        self.assertEqual(m.filesize, 87918)
        self.assertEqual(m.tags, [])
        self.assertEqual(m.size, 103950)

        match = {
            "backlinks": [],
            "domain": "",
            "score": 0,
            "format": "",
            "overlay": "",
            "image_url": "image_url",
            "filesize": 87918,
            "contributor": True,
            "size": 103950,
        }
        m = Match._from_dict(match)
        self.assertEqual(len(m.backlinks), 0)
        self.assertEqual(m.domain, "")
        self.assertEqual(m.score, 0)
        self.assertEqual(m.format, "")
        self.assertEqual(m.overlay, "")
        self.assertEqual(m.height, None)
        self.assertEqual(m.width, None)
        self.assertEqual(m.image_url, "image_url")
        self.assertEqual(m.filesize, 87918)
        self.assertEqual(m.tags, [])
        self.assertEqual(m.size, 103950)

    def test_tineye_response(self):
        """ Test TinEyeAPI.TinEyeResponse object. """

        matches = {
            "results": {
                "matches": [
                    {
                        "backlinks": [],
                        "domain": "",
                        "score": 89.36,
                        "format": "",
                        "overlay": "",
                        "height": 297,
                        "width": 350,
                        "image_url": "",
                        "filesize": 87918,
                        "contributor": True,
                        "size": 103950,
                    }
                ]
            }
        }
        r = TinEyeResponse._from_dict(matches)
        self.assertEqual(
            repr(r),
            'TinEyeResponse(matches=[Match(image_url="", score=89.36, width=350, height=297)], stats={})',
        ),
        self.assertEqual(len(r.matches), 1)
        self.assertEqual(r.matches[0].height, 297)
        self.assertEqual(r.matches[0].width, 350)

        matches = {
            "results": {
                "matches": [
                    {
                        "backlinks": [],
                        "domain": "",
                        "score": 89.36,
                        "format": "",
                        "overlay": "",
                        "height": 297,
                        "width": 350,
                        "image_url": "",
                        "filesize": 87918,
                        "tags": [],
                        "size": 103950,
                    },
                    {
                        "backlinks": [],
                        "domain": "",
                        "score": 50.4,
                        "format": "",
                        "overlay": "",
                        "height": 200,
                        "width": 300,
                        "image_url": "",
                        "filesize": 87918,
                        "tags": [],
                        "size": 103950,
                    },
                ]
            }
        }
        r = TinEyeResponse._from_dict(matches)
        self.assertEqual(len(r.matches), 2)
        self.assertEqual(r.matches[1].height, 200)
        self.assertEqual(r.matches[1].width, 300)

        matches = {"results": {"matches": []}}
        r = TinEyeResponse._from_dict(matches)
        self.assertEqual(len(r.matches), 0)

    def test_calls(self):
        """ Test methods with API sandbox. """

        # Test search_url with sandbox
        response = self.api.search_url(b"https://tineye.com/images/meloncat.jpg")
        self.assertEqual(len(response.matches), 100)
        self.assertTrue(response.stats["total_results"] > 1000)

        response = self.api.search_url(
            "https://tineye.com/images/meloncat.jpg", limit=10
        )
        self.assertEqual(len(response.matches), 10)
        self.assertTrue(response.stats["total_results"] > 1000)

        # Test search_data with sandbox
        filename = "test/images/meloncat.jpg"
        data = ""
        with open(filename, "rb") as fp:
            data = fp.read()
        response = self.api.search_data(data)
        self.assertEqual(len(response.matches), 100)
        self.assertTrue(response.stats["total_results"] > 1000)

        response = self.api.search_data(data, limit=10)
        self.assertEqual(len(response.matches), 10)
        self.assertTrue(response.stats["total_results"] > 1000)

        # Test remaining_searches with sandbox
        remaining_searches = self.api.remaining_searches()
        self.assertEqual(remaining_searches["total_remaining_searches"], 5000)
        self.assertTrue(remaining_searches["bundles"][0]["remaining_searches"], 5000)
        self.assertTrue("start_date" in remaining_searches["bundles"][0])
        self.assertTrue("expire_date" in remaining_searches["bundles"][0])
        self.assertTrue(
            isinstance(remaining_searches["bundles"][0]["start_date"], datetime)
        )
        self.assertTrue(
            isinstance(remaining_searches["bundles"][0]["expire_date"], datetime)
        )

        # Test image_count with sandbox
        image_count = self.api.image_count()
        self.assertTrue(image_count > 10000000000)

    def test_total_results_in_response(self):
        """ Test if TinEyeAPI.TinEyeResponse contains total_results. """

        response = {"results": {"matches": []}, "stats": {"total_results": 123}}
        r = TinEyeResponse._from_dict(response)
        self.assertEqual(r.stats["total_results"], 123)


test/test_api_request.py
# -*- coding: utf-8 -*-

"""
test_api_request.py

Test APIRequest class.

Copyright (c) 2021 TinEye. All rights reserved worldwide.
"""

import unittest

from pytineye.api_request import APIRequest
from pytineye.exceptions import APIRequestError


class TestAPIRequest(unittest.TestCase):
    """ Test APIRequest class. """

    def setUp(self):
        self.request = APIRequest(
            "https://api.tineye.com/rest/",
            "LCkn,2K7osVwkX95K4Oy",
            "6mm60lsCNIB,FwOWjJqA80QZHh9BMwc-ber4u=t^",
        )

    def tearDown(self):
        pass

    def test_generate_nonce(self):
        """ Test APIRequest._generate_nonce(). """

        allowable_chars = (
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRTSUVWXYZ0123456789-_=.,*^"
        )

        self.assertRaises(
            APIRequestError, lambda: self.request._generate_nonce(nonce_length=-1)
        )
        self.assertRaises(
            APIRequestError, lambda: self.request._generate_nonce(nonce_length=0)
        )
        self.assertRaises(
            APIRequestError, lambda: self.request._generate_nonce(nonce_length=23)
        )

        try:
            self.request._generate_nonce(nonce_length=0)
        except APIRequestError as e:
            self.assertEqual(
                e.args[0], "Nonce length must be an int between 24 and 255 chars"
            )

        try:
            self.request._generate_nonce(nonce_length="character")
        except APIRequestError as e:
            self.assertEqual(
                e.args[0], "Nonce length must be an int between 24 and 255 chars"
            )

        nonce = self.request._generate_nonce(nonce_length=24)
        self.assertEqual(len(nonce), 24)

        for char in nonce:
            self.assertTrue(char in allowable_chars)

        nonce = self.request._generate_nonce(nonce_length=36)
        self.assertEqual(len(nonce), 36)

        for char in nonce:
            self.assertTrue(char in allowable_chars)

    def test_generate_hmac_signature(self):
        """
        Test APIRequest._generate_get_hmac_signature(),
        TestAPIRequest._generate_post_hmac_signature(), and
        TestAPIRequest._generate_hmac_signature().
        """

        signature = self.request._generate_hmac_signature("")
        self.assertEqual(
            signature,
            "70eaf20393eb0605270226d522dfd19d281e64ea513c271c9e6febdd826a5f7c",
        )

        signature = self.request._generate_hmac_signature(" ")
        self.assertEqual(
            signature,
            "5f5f7e12a03ce8ac37ef8e98ac77dad52f0d380752ab4928c9e828949eb4e721",
        )

        signature = self.request._generate_hmac_signature(
            "this is a message to convert"
        )
        self.assertEqual(
            signature,
            "a5cb1334a32b5bb43755393be47aea8182a8557097e0f70803dc85b8d4d18562",
        )

        signature = self.request._generate_hmac_signature(
            "this is another message to convert"
        )
        self.assertEqual(
            signature,
            "ce7c686d266ebeb4e8e9dc1d1cb3d88b72ed8d4f3be41096aa3c397e45ade372",
        )

        nonce = "a_nonce"
        date = 1347910390
        signature = self.request._generate_get_hmac_signature(
            "image_count", nonce, date
        )
        self.assertEqual(
            signature,
            "dfc8b4735ab41c907059b473727bd500645f161398133db1c2ebdf276221d681",
        )

        signature = self.request._generate_get_hmac_signature(
            "remaining_searches", nonce, date, request_params={"param_1": "value"}
        )
        self.assertEqual(
            signature,
            "3b055b83d6e32f1604328306a51ba9b243f52986ba0458a7915d864d00d2f04e",
        )

        boundary = "--boundary!"
        signature = self.request._generate_post_hmac_signature(
            "search", boundary, nonce, date, filename="file"
        )
        self.assertEqual(
            signature,
            "26e789aaaf7e3f0b3c2eea15ba385a1e22da6cf93698ec98ea61627c84e35a5e",
        )

        signature = self.request._generate_post_hmac_signature(
            "search",
            boundary,
            nonce,
            date,
            filename="file",
            request_params={"param_1": "value"},
        )
        self.assertEqual(
            signature,
            "79232caabb7433561142f465cb2e645197bbc5c56a3d3832884d8e24ed128e33",
        )

    def test_sort_params(self):
        """ Test APIRequest._sort_params(). """

        request_params = {}
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "")

        request_params = {"a_param": "value_1"}
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "a_param=value_1")

        request_params = {"a_param": "value_1", "b_param": "value_2"}
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "a_param=value_1&b_param=value_2")

        request_params = {
            "param_1": "value_1",
            "a_param": "value_2",
            "b_param": "value_3",
        }
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(
            sorted_params, "a_param=value_2&b_param=value_3&param_1=value_1"
        )

        request_params = {
            "api_key": "a_key",
            "param_1": "value_1",
            "a_param": "value_2",
            "b_param": "value_3",
        }
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(
            sorted_params, "a_param=value_2&b_param=value_3&param_1=value_1"
        )

        request_params = {
            "api_key": "a_key",
            "param_1": "value_1",
            "api_sig": "a_sig",
            "a_param": "value_2",
        }
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "a_param=value_2&param_1=value_1")

        request_params = {
            "api_key": "a_key",
            "date": "date",
            "api_sig": "a_sig",
            "nonce": "nonce",
        }
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "")

        request_params = {"image_url": "valu$_1"}
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "image_url=valu%24_1")

        request_params = {"image_url": "CAPS!??!?!"}
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "image_url=caps%21%3f%3f%21%3f%21")

        request_params = {"image_url": "CAPS%21%3f%3f%21%3f%21"}
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "image_url=caps%2521%253f%253f%2521%253f%2521")

        request_params = {"Image_Url": "CAPS%21%3f%3f%21%3f%21"}
        sorted_params = self.request._sort_params(request_params)
        self.assertEqual(sorted_params, "image_url=caps%2521%253f%253f%2521%253f%2521")

    def test_request_url(self):
        """ Test APIRequest._request_url(). """

        nonce = "a_nonce"
        date = 1345821763

        url = self.request._request_url(
            "search", nonce, date, "api_sig", {"a_param": "value_1"}
        )
        self.assertEqual(
            url,
            (
                "https://api.tineye.com/rest/search/?api_key=LCkn,2K7osVwkX95K4Oy&date=1345821763"
                "&nonce=a_nonce&api_sig=api_sig&a_param=value_1"
            ),
        )

        url = self.request._request_url(
            "search",
            nonce,
            date,
            "api_sig",
            {"param_1": "value_1", "a_param": "value_2", "b_param": "value_3"},
        )
        self.assertEqual(
            url,
            (
                "https://api.tineye.com/rest/search/?api_key=LCkn,2K7osVwkX95K4Oy&date=1345821763"
                "&nonce=a_nonce&api_sig=api_sig&a_param=value_2&b_param=value_3&param_1=value_1"
            ),
        )

        url = self.request._request_url(
            "search", nonce, date, "api_sig", {"image_url": "CAPS?!!?"}
        )
        self.assertEqual(
            url,
            (
                "https://api.tineye.com/rest/search/?api_key=LCkn,2K7osVwkX95K4Oy&date=1345821763"
                "&nonce=a_nonce&api_sig=api_sig&image_url=CAPS%3F%21%21%3F"
            ),
        )


generated_docs/searchindex.js
Search.setIndex({docnames:["index"],envversion:{"sphinx.domains.c":2,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":4,"sphinx.domains.index":1,"sphinx.domains.javascript":2,"sphinx.domains.math":2,"sphinx.domains.python":3,"sphinx.domains.rst":2,"sphinx.domains.std":2,sphinx:56},filenames:["index.rst"],objects:{"pytineye.TinEyeAPIRequest":{image_count:[0,1,1,""],remaining_searches:[0,1,1,""],search_data:[0,1,1,""],search_url:[0,1,1,""]},pytineye:{Backlink:[0,0,1,""],Match:[0,0,1,""],TinEyeAPIRequest:[0,0,1,""],TinEyeResponse:[0,0,1,""]}},objnames:{"0":["py","class","Python class"],"1":["py","method","Python method"]},objtypes:{"0":"py:class","1":"py:method"},terms:{"0":0,"10":0,"100":0,"12":0,"14":0,"2019":0,"2021":0,"2023":0,"23":0,"24":0,"3":0,"45":0,"46":0,"48560880094":0,"50000":0,"52":0,"57892":0,"7892":0,"9":0,"byte":0,"class":0,"default":0,"import":0,"return":0,about:0,activ:0,an:0,api_kei:0,api_url:0,area:0,argument:0,asc:0,ascend:0,attribut:0,belong:0,block:0,bundl:0,close:0,collect:0,com:0,commun:0,connect:0,count:0,crawl:0,crawl_dat:0,criteria:0,current:0,data:0,date:0,datetim:0,debug:0,desc:0,descend:0,dictionari:0,domain:0,eas:0,end:0,establish:0,expire_d:0,extra:0,file:0,files:0,format:0,found:0,fp:0,from:0,get:0,have:0,height:0,how:0,http:0,imag:0,image_count:0,image_format:0,image_url:0,index:0,indic:0,inform:0,intend:0,jpg:0,kwarg:0,left:0,limit:0,link:0,list:0,meloncat:0,must:0,none:0,number:0,object:0,offset:0,open:0,order:0,origin:0,overlai:0,pass:0,perform:0,pixel:0,point:0,rb:0,read:0,remain:0,remaining_search:0,repres:0,respons:0,rest:0,result:0,score:0,search:0,search_data:0,search_url:0,server:0,size:0,sort:0,start:0,start_dat:0,stat:0,stock:0,tag:0,thi:0,time:0,total_remaining_search:0,url:0,urlencod:0,us:0,wa:0,websit:0,whether:0,width:0,you:0,your:0,your_api_kei:0},titles:["<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">pytineye</span></code> - Python client for the TinEye API"],titleterms:{api:0,backlink:0,client:0,match:0,python:0,pytiney:0,tiney:0,tineyeapirequest:0,tineyerespons:0}})

pytineye/api.py
# -*- coding: utf-8 -*-

"""
api.py

Python library to ease communication with the TinEye API server.

Copyright (c) 2021 TinEye. All rights reserved worldwide.
"""

import http.client
import json
import time
from datetime import datetime

import certifi
import urllib3

from .exceptions import TinEyeAPIError

try:
    from urllib.parse import urljoin
except ImportError:
    from urlparse import urljoin


class TinEyeResponse(object):
    """
    Represents a response from the API.

    Attributes:

    - `matches`, a list of Match objects.
    - `stats`, stats for this search.
    """

    def __init__(self, matches, stats):
        self.matches = matches
        self.stats = stats

    def __repr__(self):
        return "%s(matches=%s, stats=%s)" % (
            self.__class__.__name__,
            self.matches,
            self.stats,
        )

    @staticmethod
    def _from_dict(result_json):
        """
        Takes parsed JSON from the API server and turns it into a
        TinEyeResponse object.

        - `result_json`, the parsed JSON object.

        Returns: a TinEyeResponse object.
        """

        if not isinstance(result_json, dict):
            raise TinEyeAPIError("500", ["Please pass in a dictionary to _from_dict()"])

        matches = []
        stats = {}

        if "results" in result_json:
            results = result_json["results"]
            if "matches" in results:
                for m in results.get("matches"):
                    match = Match._from_dict(m)
                    matches.append(match)
        if "stats" in result_json:
            stats = result_json["stats"]

        return TinEyeResponse(
            matches=matches,
            stats=stats,
        )


class Match(object):
    """
    Represents an image match.

    Attributes:

    - `image_url`, link to the result image.
    - `domain`, domain this result was found on.
    - `score`, a number (0 to 100) that indicates how closely the images match.
    - `width`, image width in pixels.
    - `height`, image height in pixels.
    - `size`, image area in pixels.
    - `format`, image format.
    - `filesize`, image size in bytes.
    - `overlay`, overlay URL.
    - `backlinks`, a list of Backlink objects pointing to
      the original websites and image URLs.
    - `tags`, whether this match belongs to a collection or stock domain.
    """

    def __init__(
        self,
        image_url,
        domain,
        score,
        width,
        height,
        size,
        image_format,
        filesize,
        overlay,
        tags=None,
        backlinks=None,
    ):
        self.image_url = image_url
        self.domain = domain
        self.score = score
        self.width = width
        self.height = height
        self.size = size
        self.format = image_format
        self.filesize = filesize
        self.overlay = overlay
        self.backlinks = backlinks
        if not tags:
            self.tags = []
        else:
            self.tags = tags

    def __repr__(self):
        return '%s(image_url="%s", score=%.2f, width=%i, height=%i)' % (
            self.__class__.__name__,
            self.image_url,
            self.score,
            self.width,
            self.height,
        )

    @staticmethod
    def _from_dict(match_json):
        """
        Takes parsed JSON from the API server and turns it into a
        Match object.

        - `match_json`, the parsed JSON object.

        Returns: a Match object.
        """

        if not isinstance(match_json, dict):
            raise TinEyeAPIError("500", ["Please pass in a dictionary to _from_dict()"])

        backlinks = []
        if "backlinks" in match_json:
            for b in match_json["backlinks"]:
                backlinks.append(Backlink._from_dict(b))
        match = Match(
            image_url=match_json.get("image_url"),
            domain=match_json.get("domain"),
            score=match_json.get("score"),
            width=match_json.get("width"),
            height=match_json.get("height"),
            size=match_json.get("size"),
            image_format=match_json.get("format"),
            filesize=match_json.get("filesize"),
            overlay=match_json.get("overlay"),
            tags=match_json.get("tags"),
            backlinks=backlinks,
        )
        return match


class Backlink(object):
    """
    Represents a backlink for an image.

    Attributes:

    - `url`, the image URL to the image.
    - `backlink`, the original website URL.
    - `crawl_date`, the date the image was crawled.
    """

    def __init__(self, url=None, backlink=None, crawl_date=None):
        self.url = url
        self.backlink = backlink
        self.crawl_date = crawl_date

    def __repr__(self):
        return '%s(url="%s", backlink="%s", crawl_date=%s)' % (
            self.__class__.__name__,
            self.url,
            self.backlink,
            str(self.crawl_date),
        )

    @staticmethod
    def _from_dict(backlink_json):
        """
        Takes parsed JSON from the API server and turns it into a
        Backlink object.

        - `backlink_json`, the parsed JSON object.

        Returns: a Backlink object.
        """

        if not isinstance(backlink_json, dict):
            raise TinEyeAPIError("500", ["Please pass in a dictionary to _from_dict()"])

        crawl_date = datetime.min
        if backlink_json.get("crawl_date"):
            crawl_date = time.strptime(backlink_json.get("crawl_date"), "%Y-%m-%d")
            crawl_date = datetime(*crawl_date[:6])
        return Backlink(
            url=backlink_json.get("url"),
            backlink=backlink_json.get("backlink"),
            crawl_date=crawl_date,
        )


class TinEyeAPIRequest(object):
    """
    Class to ease communication with the TinEye API server.

    Establish a connection to the API:

        >>> from pytineye import TinEyeAPIRequest
        >>> api = TinEyeAPIRequest(
        ...     api_url='https://api.tineye.com/rest/',
        ...     api_key='your_api_key'
        ... )

    Searching for an image using an image URL:

        >>> api.search_url(url='https://tineye.com/images/meloncat.jpg')
        TinEyeResponse(...)

    Searching for an image using image data:

        >>> fp = open('meloncat.jpg', 'rb')
        >>> data = fp.read()
        >>> api.search_data(data=data)
        TinEyeResponse(...)
        >>> fp.close()

    Getting information about your search bundle:

        >>> api.remaining_searches()
        {'bundles': [{'expire_date': datetime.datetime(2023, 3, 9, 14, 9, 12),
           'remaining_searches': 7892,
           'start_date': datetime.datetime(2021, 3, 10, 14, 9, 12)},
          {'expire_date': datetime.datetime(2019, 3, 23, 9, 52, 46),
           'remaining_searches': 50000,
           'start_date': datetime.datetime(2021, 3, 24, 9, 52, 45)}],
         'total_remaining_searches': 57892}

    Getting an image count:

        >>> api.image_count()
        48560880094

    """

    def __init__(
        self,
        api_url="https://api.tineye.com/rest/",
        api_key="",
        connect_timeout=15.0,
        read_timeout=60.0,
    ):
        self.http_pool = urllib3.PoolManager(
            cert_reqs="CERT_REQUIRED",
            ca_certs=certifi.where(),
            timeout=urllib3.Timeout(connect=connect_timeout, read=read_timeout),
        )
        self.api_url = api_url
        self.api_key = api_key

    def _request(self, method, params=None, **kwargs):
        """
        Send request to API and process results.

        - `method`, API method to call.
        - `params`, dictionary of fields to send to the API call.
        - `image_file`, tuple containing info (filename, data) about image to send.

        Returns: a JSON parsed object.
        """

        # Pass in any extra keyword arguments as parameters to the API call
        if not params:
            params = {}
        params.update(kwargs)

        headers = {"x-api-key": self.api_key}
        api_url_search = urljoin(self.api_url, "/rest/%s/" % method)

        try:
            obj = None
            response = None

            # If an image file was provided, send a POST request, else send a GET request
            if "image_upload" not in params:
                response = self.http_pool.request(
                    method="GET", url=api_url_search, fields=params, headers=headers
                )
            else:
                response = self.http_pool.request_encode_body(
                    method="POST",
                    url=api_url_search,
                    fields=params,
                    headers=headers,
                )
            # Parse the JSON into a Python object
            obj = json.loads(response.data.decode("utf-8"))

        except ValueError as e:
            raise TinEyeAPIError("500", ["Could not decode JSON: %s" % e])

        # Check the result of the API call
        if response.status != http.client.OK or obj.get("code") != http.client.OK:
            raise TinEyeAPIError(obj["code"], obj.get("messages"))

        return obj

    def search_url(
        self, url, offset=0, limit=100, sort="score", order="desc", **kwargs
    ):
        """
        Perform searches on the TinEye index using an image URL.

        - `url`, the URL of the image that will be searched for, must be urlencoded.
        - `offset`, offset of results from the start, defaults to 0.
        - `limit`, number of results to return, defaults to 100.
        - `sort`, sort results by score, file size (size), or crawl date (crawl_date),
          defaults to descending (desc).
        - `order`, sort results by ascending (asc) or descending criteria.
        - `kwargs`, to pass extra arguments intended for debugging.

        Returns: a TinEye Response object.
        """

        params = {
            "image_url": url,
            "offset": offset,
            "limit": limit,
            "sort": sort,
            "order": order,
        }
        obj = self._request("search", params, **kwargs)

        return TinEyeResponse._from_dict(obj)

    def search_data(
        self, data, offset=0, limit=100, sort="score", order="desc", **kwargs
    ):
        """
        Perform searches on the TinEye index using image data.

        - `data`, image data to use for searching.
        - `offset`, offset of results from the start, defaults to 0.
        - `limit`, number of results to return, defaults to 100.
        - `sort`, sort results by score, file size (size), or crawl date (crawl_date),
          defaults to descending (desc).
        - `order`, sort results by ascending (asc) or descending criteria.
        - `kwargs`, to pass extra arguments intended for debugging.

        Returns: a TinEye Response object.
        """

        params = {
            "image_upload": ("image.jpg", data),
            "offset": offset,
            "limit": limit,
            "sort": sort,
            "order": order,
        }
        obj = self._request("search", params=params, **kwargs)

        return TinEyeResponse._from_dict(obj)

    def remaining_searches(self, **kwargs):
        """
        Lists the number of searches you have left in your current active block.

        - `kwargs`, to pass extra arguments intended for debugging.

        Returns: a dictionary with remaining searches, start time and end time of block.
        """

        bundle_list = []

        obj = self._request("remaining_searches", **kwargs)

        results = obj.get("results")

        for bundle in results.get("bundles"):

            start_date = time.strptime(bundle.get("start_date"), "%Y-%m-%d %X UTC")
            start_date = datetime(*start_date[:6])
            expire_date = time.strptime(bundle.get("expire_date"), "%Y-%m-%d %X UTC")
            expire_date = datetime(*expire_date[:6])

            bundle_list.append(
                {
                    "remaining_searches": bundle.get("remaining_searches"),
                    "start_date": start_date,
                    "expire_date": expire_date,
                }
            )

        return {
            "bundles": bundle_list,
            "total_remaining_searches": results.get("total_remaining_searches"),
        }

    def image_count(self, **kwargs):
        """
        Lists the number of indexed images on TinEye.

        - `kwargs`, to pass extra arguments intended for debugging.

        Returns: TinEye image count.
        """

        obj = self._request("image_count", **kwargs)
        return obj.get("results")
